compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -460,26 +478,42 @@ impl<'a, 'b, 'tcx> TypeFolder<'tcx> for AssocTypeNormalizer<'a, 'b, 'tcx> {
 }
 
 pub struct BoundVarReplacer<'me, 'tcx> {
-    pub infcx: &'me InferCtxt<'me, 'tcx>,
-    pub mapped_regions: BTreeMap<ty::PlaceholderRegion, ty::BoundRegion>,
-    pub mapped_types: BTreeMap<ty::PlaceholderType, ty::BoundTy>,
-    pub mapped_consts: BTreeMap<ty::PlaceholderConst<'tcx>, ty::BoundVar>,
-    pub universes: BTreeMap<ty::DebruijnIndex, ty::UniverseIndex>,
-    pub universes_inverse: BTreeMap<ty::UniverseIndex, ty::DebruijnIndex>,
-    pub current_index: ty::DebruijnIndex,
+    infcx: &'me InferCtxt<'me, 'tcx>,
+    // These three maps track the bound variable that were replaced by placeholders. It might be
+    // nice to remove these since we already have the `kind` in the placeholder; we really just need
+    // the `var` (but we *could* bring that into scope if we were to track them as we pass them).
+    mapped_regions: BTreeMap<ty::PlaceholderRegion, ty::BoundRegion>,
+    mapped_types: BTreeMap<ty::PlaceholderType, ty::BoundTy>,
+    mapped_consts: BTreeMap<ty::PlaceholderConst<'tcx>, ty::BoundVar>,
+    // The current depth relative to *this* folding, *not* the entire normalization. In other words,
+    // the depth of binders we've passed here.
+    current_index: ty::DebruijnIndex,
+    // The `UniverseIndex` of the binding levels above us. These are optional, since we are lazy:
+    // we don't actually create a universe until we see a bound var we have to replace.
+    universe_indices: &'me mut Vec<Option<ty::UniverseIndex>>,
+    // FIXME: So, this is a less-than-ideal solution to a problem we want to solve eventually. Ideally, we
+    // shouldn't need to worry about bound vars for which we haven't passed (`self.current_index`)
+    // or that we don't explicitly know about (`self.universe_indices`). This is true for
+    // `AssocTypeNormalizer` but not `QueryNormalizer` currently. When we can always know about
+    // any binding levels above us, we can remove this. (The alternative would be
+    // `outer_exclusive_binder`, but that only exists on `Ty`. Otherwise, we would have to visit
+    // through the `T`, which we specifically want to avoid not being lazy.)
+    failed: bool,
```

Comment:
As discussed on Zulip, replace this with an assertion (or an `if`) to prevent us from getting into this code in the first place.